ledger/error/generator/lib/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala

```diff
@@ -8,239 +8,229 @@ import java.lang.reflect.Modifier
-          Explanation(parseAnnotationValue(annotation.tree))
-        }
+  private def parseErrorGroupAnnotations(errorGroup: ErrorGroup): ErrorGroupAnnotations = {
+    val mirroredType = runtimeMirror.reflect(errorGroup)
+    val annotations = mirroredType.symbol.annotations
+    val explanation = new SettableOnce[Explanation]
+    annotations.foreach { annotation =>
+      getAnnotationTypeName(annotation) match {
+        case ExplanationTypeName =>
+          explanation.set(Explanation(parseAnnotationValue(annotation.tree)), ExplanationTypeName)
+        case otherAnnotationTypeName =>
+          throw new IllegalStateException(
+            s"Unexpected annotation of type: ${otherAnnotationTypeName}"
+          )
+      }
+    }
+    ErrorGroupAnnotations(
+      explanation = explanation.get
     )
   }
 
-  private case class GetAnnotationsState(
-      deprecation: Option[DeprecatedDocs],
-      explanation: Option[Explanation],
-      resolution: Option[Resolution],
-  )
+  private def findInstancesOf[T: ru.TypeTag](packagePrefixes: Array[String]): Seq[T] =
+    new Reflections(packagePrefixes)
+      .getSubTypesOf(runtimeMirror.runtimeClass(ru.typeOf[T]))
+      .asScala
+      .view
+      .collect {
+        case clazz if !Modifier.isAbstract(clazz.getModifiers) =>
+          clazz.getDeclaredField("MODULE$").get(clazz).asInstanceOf[T]
+      }
+      .toSeq
 
-  private def getErrorAnnotations(
-      annotations: Seq[ru.Annotation]
-  ): ErrorDocumentationAnnotations = {
-
-    def update(
-        state: GetAnnotationsState,
-        updatedDeprecation: Option[String] = None,
-        updatedExplanation: Option[String] = None,
-        updatedResolution: Option[String] = None,
-    ): GetAnnotationsState = {
-
-      def updateString(
-          existing: Option[String],
-          updated: Option[String],
-          designation: String,
-      ): Option[String] =
-        updated.fold(ifEmpty = existing) { value: String =>
-          if (existing.isDefined)
-            sys.error(s"Multiple $designation annotations detected")
-          else
-            Some(value)
-        }
-
-      val existingDeprecation = state.deprecation
-      val updatedDeprecationString =
-        updateString(existingDeprecation.map(_.deprecation), updatedDeprecation, "deprecation")
-      val existingExplanation = state.explanation
-      val updatedExplanationString =
-        updateString(existingExplanation.map(_.explanation), updatedExplanation, "explanation")
-      val existingResolution = state.resolution
-      val updatedResolutionString =
-        updateString(existingResolution.map(_.resolution), updatedResolution, "resolution")
-      GetAnnotationsState(
-        updatedDeprecationString.map(DeprecatedDocs),
-        updatedExplanationString.map(Explanation),
-        updatedResolutionString.map(Resolution),
-      )
-    }
+  private def simpleClassName(any: Any): String =
+    any.getClass.getSimpleName.replace("$", "")
```

Comment:
Watch out for this, `.getSimpleName` throws for inner class definitions in Scala.

Reply:
Only moved this code around so I'd like to avoid investing more time in it :(